service/tablet_allocator.cc

```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
         }
 
+        // Below section ensures we respect the _tablets_per_shard_goal.
+        //
+        // It will scale down target_tablet_count for all tables so that
+        // the average number of tablets per shard in each DC does not exceed _tablets_per_shard_goal.
+        //
+        // The impact of table's tablet count on average per-shard tablet replica count
+        // is different in each DC because replication factors are different in each DC.
+        //
+        // The algorithm works like this:
+        // Compute average tablet replica count per-shard in each DC,
+        // determine if per-shard goal is exceeded in that DC,
+        // compute scale factor by which tablet count should be multiplied so that the goal
+        // is not exceeded in that DC.
+        // Take the smallest scale factor among all DCs, which ensures that no DC is overloaded.
+        //
+        // We align tablet counts to the nearest power of 2 post-scaling, which
+        // means that scaling may not be effective and in the worst case we may overshoot the goal by
+        // a factor of 2. This is acceptable since the goal is a soft limit and not a hard constraint.
+        // Scaling post-alignment would be problematic. If we scale down all tables fairly, we undershoot the goal
+        // by a factor of 2 in the worst case. If we choose a subset of tables to scale down by a factor of 2 then
+        // we have a problem of making sure that the choice is stable across scheduler invocations to avoid
+        // oscillations of decisions.
+
+        std::unordered_map<table_id, double> table_scaling;
+
+        std::unordered_map<sstring, unsigned> shards_per_dc;
+        _tm->get_topology().for_each_node([&] (const node& n) {
+            if (n.is_normal() && _tm->is_normal_token_owner(n.host_id())) {
+                shards_per_dc[n.dc_rack().dc] += n.get_shard_count();
+            }
+        });
+
+        for (auto&& [dc, shard_count] : shards_per_dc) {
+            double old_avg_tablets_per_shard = 0;
+            double avg_tablets_per_shard = 0;
+
+            for (auto&& [table, table_plan] : plan.tables) {
+                auto* rs = rs_by_table[table];
+                if (!rs) {
+                    continue;
+                }
+
+                auto rf = rs->get_replication_factor(dc);
+                auto get_avg_tablets_per_shard = [&] (size_t tablet_count) {
+                    return double(tablet_count) * rf / shard_count;
+                };
+
+                auto old_tablets_per_shard = get_avg_tablets_per_shard(table_plan.current_tablet_count);
+                old_avg_tablets_per_shard += old_tablets_per_shard;
+                lblogger.debug("old_avg_tablets_per_shard [dc={}, table={}]: {:.3f}", dc, table, old_tablets_per_shard);
+
+                auto tablets_per_shard = get_avg_tablets_per_shard(table_plan.target_tablet_count);
+                avg_tablets_per_shard += tablets_per_shard;
```

Comment:
I don't see where that's normalized by the number of tables.
Or did I miss something?

Reply:
Why should we normalize by the number of tables? We're computing per-shard load, because the limit is for tablets per shard. Number of tables doesn't change anything. We can have one table with 100 tablets per shard, or 100 tables with 1 tablet per shard.

Reply:
ok, got it.  It is the per-shard tablet count goal we're interested in here, not per table per_shard_tablet_count.